features/ALB001_Alignment-in-spatial-structure.feature

```diff
@@ -0,0 +1,9 @@
+@implementer-agreement
+@ALB
+Feature: ALB001_Alignment in spatial structure
+
+  Scenario: Agreement on each IfcAlignment being contained in an IfcSite
+
+      Given A file with Schema Identifier "IFC4X3"
+      And An IfcAlignment
+      Then Each IfcAlignment must be directly contained in IfcSite
```

Comment:
What does it look like if something is `indirectly` contained?

Reply:
The intention was to have `Indirectly` take into account spatial relationships. Let's provide some examples.

Directly:
IfcAlignment > IfcSite 
IfcAlignment connected to IfcSite via the relationship IFCRELCONTAINEDINSPATIALSTRUCTURE

Indirectly: 
IfcAlignment > SpacialElement1 > SpacialElement2 > IfcSite
IfcAlignment connected to SpacialElement1 via the relationship IFCRELCONTAINEDINSPATIALSTRUCTURE
SpacialElement1 to SpacialElement2  via the relationship IFCRELAGGREGATES
SpacialElement2  to IfcSite via the relationship IFCRELAGGREGATES

Maybe it is worthwhile to try different wording. Perhaps the following?
`Each IfcAlignment must be directly or through spatial relationships contained in IfcSite`

@pjanck @aothms @evandroAlfieri

Reply:
Maybe it helps to think of this as a regular expression or sparql property path. E.g what's the pattern of relationships that can be followed in order to be valid.

```
elem -> ContainedIn -> Site
elem -> Decomposing -> Assembly -> ContainedIn -> Site
elem -> Decomposing -> Assembly -> ContainedIn -> Something -> Decomposing -> Site
```

We could write that as something like:

    elem -> Decomposing * -> ContainedIn -> Decomposing * -> Site
    
Where * means zero or more repetitions of the preceeding relationship.

I don't think multiple ContainedIn relationships would be allowed actually in the IFC spec, so "directly or through spatial relationships" maybe better be "directly or through decomposition relationships"

The question is also do we want to make this more explicit in Gherkin (`following a pattern of one or more RelX exactly one RelY zero ore more RelZ`) or do we want to keep it something descriptive (e.g `indirectly`) matched in code? The ever remaining tension between explicit IFC constructs in rules (nice for people familiar with them, less so for the people that predominantly care about business logic).

If we want to make this more explicit in Gherkin we can rely on Behave to do the course matching for us but in the implementation use a custom grammar to more precisely match against the semantics of the predicate/sentence fragment. Something like this:

```
>>> from pyparsing import *
>>> zom = CaselessKeyword("zero or more")
>>> eo = CaselessKeyword("exactly one")
>>> entity = Word(alphas)('entity')
>>> grammar = OneOrMore((zom | eo) + entity)
>>> grammar.parseString("zero or more IfcRelX exactly one IfcRelY zero or more IfcRelZ")
ParseResults(['zero or more', 'IfcRelX', 'exactly one', 'IfcRelY', 'zero or more', 'IfcRelZ'], {'entity': 'IfcRelZ'})
```